File: gateway/src/policy.rs
//! Policy model and evaluation for the MCP Bastion gateway.
//!
//! Policies are expressed in a tiny, line-oriented, dependency-free format so
//! the whole gateway can rely solely on the Rust standard library. Every
//! directive is `key = value` or `key value` on its own line. Lines beginning
//! with `#` are comments; blank lines are ignored.
//!
//! Supported directives (see `docs/POLICY.md` for the authoritative reference):
//!
//! ```text
//! default            = deny            # deny | allow  (fallback decision)
//! allow_tool         = read_file       # may be repeated
//! deny_tool          = shell.exec      # may be repeated; deny wins over allow
//! redact_arg         = token           # redact this argument key in tools/call
//! redact_arg         = *password*      # `*` wildcards are supported
//! max_bytes          = 65536           # reject messages larger than this
//! max_depth          = 32              # audit-flag messages nested deeper
//! rate_limit         = 20              # max forwarded messages per window
//! rate_window_ms     = 1000            # sliding window length
//! redaction_mask     = «redacted»      # replacement text for redacted values
//! ```
//!
//! The `allow_tool` / `deny_tool` lists match the `name` field of MCP
//! `tools/call` requests. `deny` always beats `allow`. If a tool is on neither
//! list, the `default` decision applies.

use std::collections::HashMap;

/// A fully-resolved policy.
#[derive(Debug, Clone)]
pub struct Policy {
    pub default_allow: bool,
    pub allow_tools: Vec<Pattern>,
    pub deny_tools: Vec<Pattern>,
    pub redact_args: Vec<Pattern>,
    pub max_bytes: usize,
    pub max_depth: usize,
    pub rate_limit: u32,
    pub rate_window_ms: u64,
    pub redaction_mask: String,
}

impl Default for Policy {
    fn default() -> Self {
        Policy {
            default_allow: false,
            allow_tools: Vec::new(),
            deny_tools: Vec::new(),
            redact_args: Vec::new(),
            max_bytes: 256 * 1024,
            max_depth: 64,
            rate_limit: 0, // 0 == unlimited
            rate_window_ms: 1000,
            redaction_mask: "«redacted»".to_string(),
        }
    }
}

/// A simple case-sensitive glob supporting `*` wildcards (any run of chars).
#[derive(Debug, Clone)]
pub struct Pattern {
    raw: String,
    /// Literal segments that must appear in order; `None` markers between them
    /// represent `*`. `anchored_start`/`anchored_end` say whether the pattern
    /// touches the respective boundary.
    parts: Vec<String>,
    anchored_start: bool,
    anchored_end: bool,
}

impl Pattern {
    pub fn new(raw: &str) -> Pattern {
        let anchored_start = !raw.starts_with('*');
        let anchored_end = !raw.ends_with('*');
        let parts: Vec<String> = raw
            .split('*')
            .filter(|s| !s.is_empty())
            .map(|s| s.to_string())
            .collect();
        Pattern {
            raw: raw.to_string(),
            parts,
            anchored_start,
            anchored_end,
        }
    }

    pub fn raw(&self) -> &str {
        &self.raw
    }

    /// Does this pattern match the whole of `text`?
    pub fn matches(&self, text: &str) -> bool {
        if self.parts.is_empty() {
            // Pattern was "" or "*" or "***" -> matches everything.
            return true;
        }
        let mut pos = 0usize;
        for (idx, part) in self.parts.iter().enumerate() {
            let is_first = idx == 0;
            let is_last = idx == self.parts.len() - 1;
            if is_first && self.anchored_start {
                if !text[pos..].starts_with(part.as_str()) {
                    return false;
                }
                pos += part.len();
            } else {
                match text[pos..].find(part.as_str()) {
                    Some(rel) => pos += rel + part.len(),
                    None => return false,
                }
            }
            if is_last && self.anchored_end {
                // The final literal must land exactly at the end.
                return pos == text.len();
            }
        }
        true
    }
}

/// Errors that can occur while loading a policy file.
#[derive(Debug)]
pub enum PolicyError {
    UnknownDirective {
        line: usize,
        key: String,
    },
    BadValue {
        line: usize,
        key: String,
        value: String,
    },
}

impl std::fmt::Display for PolicyError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PolicyError::UnknownDirective { line, key } => {
                write!(f, "line {line}: unknown directive '{key}'")
            }
            PolicyError::BadValue { line, key, value } => {
                write!(f, "line {line}: invalid value '{value}' for '{key}'")
            }
        }
    }
}

impl std::error::Error for PolicyError {}

impl Policy {
    /// Parse a policy from its textual representation.
    pub fn parse(text: &str) -> Result<Policy, PolicyError> {
        let mut p = Policy::default();
        for (idx, raw_line) in text.lines().enumerate() {
            let line_no = idx + 1;
            let line = strip_comment(raw_line).trim();
            if line.is_empty() {
                continue;
            }
            let (key, value) = split_directive(line);
            let key = key.trim();
            let value = value.trim();
            match key {
                "default" => match value {
                    "allow" => p.default_allow = true,
                    "deny" => p.default_allow = false,
                    other => {
                        return Err(PolicyError::BadValue {

File: console/src/policy.ts
/**
 * A read-only parser and summariser for the MCP Bastion policy format.
 *
 * This mirrors the grammar implemented authoritatively in the Rust gateway
 * (`../../gateway/src/policy.rs`) so the console can *display* and lint a policy
 * without shelling out. It intentionally does not evaluate messages — that is
 * the gateway's job — but it validates directive names and value shapes so an
 * operator can sanity-check a policy file before deploying it.
 */

export interface ParsedPolicy {
  defaultAllow: boolean;
  allowTools: string[];
  denyTools: string[];
  redactArgs: string[];
  maxBytes: number;
  maxDepth: number;
  rateLimit: number;
  rateWindowMs: number;
  redactionMask: string;
}

export interface PolicyIssue {
  readonly line: number;
  readonly severity: "error" | "warning";
  readonly message: string;
}

export interface PolicyParseResult {
  readonly policy: ParsedPolicy;
  readonly issues: readonly PolicyIssue[];
}

const KNOWN_DIRECTIVES = new Set([
  "default",
  "allow_tool",
  "deny_tool",
  "redact_arg",
  "max_bytes",
  "max_depth",
  "rate_limit",
  "rate_window_ms",
  "redaction_mask",
]);

function defaults(): ParsedPolicy {
  return {
    defaultAllow: false,
    allowTools: [],
    denyTools: [],
    redactArgs: [],
    maxBytes: 262144,
    maxDepth: 64,
    rateLimit: 0,
    rateWindowMs: 1000,
    redactionMask: "«redacted»",
  };
}

function stripComment(line: string): string {
  let inQuote = false;
  for (let i = 0; i < line.length; i++) {
    const c = line[i];
    if (c === '"') inQuote = !inQuote;
    else if (c === "#" && !inQuote) return line.slice(0, i);
  }
  return line;
}

function splitDirective(line: string): [string, string] {
  const eq = line.indexOf("=");
  if (eq >= 0) return [line.slice(0, eq).trim(), line.slice(eq + 1).trim()];
  const sp = line.search(/\s/);
  if (sp >= 0) return [line.slice(0, sp).trim(), line.slice(sp + 1).trim()];
  return [line.trim(), ""];
}

function unquote(v: string): string {
  const t = v.trim();
  if (t.length >= 2 && t.startsWith('"') && t.endsWith('"')) {
    return t.slice(1, -1);
  }
  return t;
}

function parseIntStrict(v: string): number | null {
  if (!/^\d+$/.test(v)) return null;
  const n = Number(v);
  return Number.isSafeInteger(n) ? n : null;
}

/** Parse policy text, collecting issues without throwing. */
export function parsePolicy(text: string): PolicyParseResult {
  const policy = defaults();
  const issues: PolicyIssue[] = [];
  const lines = text.split(/\r?\n/);

  for (let i = 0; i < lines.length; i++) {
    const lineNo = i + 1;
    const line = stripComment(lines[i]!).trim();
    if (line.length === 0) continue;
    const [key, value] = splitDirective(line);

    if (!KNOWN_DIRECTIVES.has(key)) {
      issues.push({
        line: lineNo,
        severity: "error",
        message: `unknown directive '${key}'`,
      });
      continue;
    }

    switch (key) {
      case "default":
        if (value === "allow") policy.defaultAllow = true;
        else if (value === "deny") policy.defaultAllow = false;
        else
          issues.push({
            line: lineNo,
            severity: "error",
            message: `default must be 'allow' or 'deny', got '${value}'`,
          });
        break;
      case "allow_tool":
        policy.allowTools.push(value);
        break;
      case "deny_tool":
        policy.denyTools.push(value);
        break;
      case "redact_arg":
        policy.redactArgs.push(value);
        break;
      case "max_bytes": {
        const n = parseIntStrict(value);
        if (n === null)
          issues.push({ line: lineNo, severity: "error", message: `max_bytes must be an integer` });
